proposals/0258-property-wrappers.md

```diff
@@ -624,106 +662,148 @@ in one of two ways:
 
-Type inference for properties with delegates involves both the type
-annotation of the original property (if present) and the delegate
-type, using the initialization of the synthesized stored property. For
-example:
+   ```swift
+   @DelayedMutable var x: Int
 
-```swift
-@Lazy var foo = 17
-// type inference as in...
-var $foo: Lazy = Lazy(initialValue: 17)
-// infers the type of 'foo' to be 'Int'
-```
+   // ... implemented as
+   var $x: DelayedMutable<Int> = DelayedMutable<Int>()
+   var x: Int { /* access via $x.value */ }
+   ```
 
-The same applies when directly initializing the delegate instance, e.g.,
+  When there are multiple, composed property wrappers, only the first (outermost) wrapper needs to have an `init()`.
 
-```swift
-@UnsafeMutablePointer(mutating: addressOfInt)
-var someInt
-// type inference as in...
-var $someInt: UnsafeMutablePointer = UnsafeMutablePointer.init(mutating: addressOfInt)
-// infers the type of 'someInt' to be 'Int'
-```
+### Type inference with property wrappers
+
+If the first property wrapper type is generic, its generic arguments must either be given explicitly in the attribute or Swift must be able to deduce them from the variable declaration. That deduction proceeds as follows:
+
+* If the variable has an initial value expression `E`, then the first wrapper type is constrained to equal the type resulting from a call to `A(initialValue: E)`, where `A` is the written type of the attribute. For example:
 
-The type of the `value` property of the property delegate type must coincide with that of the original property using that delegate type. Some examples:
+  ```swift
+  @Lazy var foo = 17
+  // type inference as in...
+  var $foo: Lazy = Lazy(initialValue: 17)
+  // infers the type of '$foo' to be 'Lazy<Int>'
+  ```
+
+  If there are multiple wrapper attributes, the argument to this call will instead be a nested call to `B(initialValue: E)` for the written type of the next attribute, and so on recursively. For example:
+  
+  ```swift
+  @A @B var bar = 42
+  // type inference as in ...
+  var $bar = A(initialValue: B(initialValue: 42))
+  // infers the type of '$bar' to be 'A<B<Int>'
+  ```
+
+* Otherwise, if the first wrapper attribute has direct initialization arguments `E...`, the outermost wrapper type is constrained to equal the type resulting from `A(E...)`, where `A` is the written type of the first attribute. Wrapper attributes after the first may not have direct initializers. For example:
+
+  ```swift
+  @UnsafeMutablePointer(mutating: addressOfInt)
+  var someInt
+  // type inference as in...
+  var $someInt: UnsafeMutablePointer = UnsafeMutablePointer.init(mutating: addressOfInt)
+  // infers the type of `$someInt` to be `UnsafeMutablePointer<Int>`
+  ```
+
+* Otherwise, if there is no initialization, and the original property has a type annotation, the type of the `value` property in the last wrapper type is constrained to equal the type annotation of the original property. For example:
```

Comment:
This constraint seems to be added unconditionally in the "In any case..." paragraph below.

Reply:
I know, but I wanted to make it very clear that this can be used to infer generic arguments for the wrapper type, too. The "In any case" below applies to everything... generic wrapper type or not.

Reply:
Oh, I was assuming that the "In any case" could also be used to infer generic arguments.  Well, I don't want to make your implementation task bigger than it already is. :)